src/update_agent/mod.rs

```diff
@@ -117,6 +117,51 @@ impl UpdateAgentState {
         self.transition_to(target);
     }
 
+    /// Record a deploy attempt in UpdateAvailable state.
+    ///
+    /// In case of persistent failure, this returns the update version which
+    /// is being abandoned.
+    fn deploy_attempted(&mut self, success: bool) -> Option<Release> {
+        // Maximum failed deploy attempts before declaring a persistent error.
+        const MAX_DEPLOY_ATTEMPTS: u8 = 12;
+
+        let (release, attempts) = match self.clone() {
+            UpdateAgentState::UpdateAvailable((r, a)) => (r, a),
+            _ => unreachable!("transition not allowed: deploy_attempted on {:?}", self,),
+        };
+
+        let persistent_err = attempts >= MAX_DEPLOY_ATTEMPTS;
+        match (success, persistent_err) {
+            (true, _) => {
+                self.update_staged(release);
+                None
+            }
+            (false, false) => {
+                self.deploy_failed(release, attempts);
+                None
+            }
+            (false, true) => {
+                self.update_abandoned();
+                Some(release)
+            }
+        }
+    }
+
+    /// Transition to the UpdateAvailable state after a deploy failure.
+    fn deploy_failed(&mut self, update: Release, attempts: u8) {
+        let failed = attempts.saturating_add(1);
+        let target = UpdateAgentState::UpdateAvailable((update, failed));
+
+        self.transition_to(target);
+    }
+
+    /// Transition to the NoNewUpdate state after persistent deploy failure.
+    fn update_abandoned(&mut self) {
+        let target = UpdateAgentState::NoNewUpdate;
+
+        self.transition_to(target);
```

Comment:
So just to check, after we go back to `NoNewUpdate` here, do we ignore the same update again, or will we retry it even if that's still the latest update when we recheck? I think we want the latter behaviour, right? That way, e.g. if the OSTree remote is down for 1h or whatever, we'll still retry it again later.

Reply:
Yes. Going back to `NoNewUpdate` clears the state, so on the next attempt we may be retrying the same update (that's the latter behavior).

If the graph changed in the meanwhile, best-edge selection will result in picking up any newer target update (that's the usual logic for choosing the freshest target among a set of valid edges).